contracts/truefi/TrueRatingAgencyV2.sol

```diff
@@ -0,0 +1,485 @@
+    }
+
+    /**
+     * @dev Internal view to convert values to 8 decimals precision
+     * @param input Value to convert to TRU precision
+     * @return output TRU amount
+     */
+    function toTrustToken(uint256 input) internal pure returns (uint256 output) {
+        output = input.div(TOKEN_PRECISION_DIFFERENCE);
+    }
+
+    /**
+     * @dev Update total TRU reward for a Loan
+     * Reward is divided proportionally based on # TRU staked
+     * chi = (TRU remaining in distributor) / (Total TRU allocated for distribution)
+     * interest = (loan APY * term * principal)
+     * R = Total Reward = (interest * chi * rewardFactor)
+     * @param id Loan ID
+     */
+    modifier calculateTotalReward(address id) {
+        if (loans[id].reward == 0) {
+            uint256 interest = ILoanToken(id).profit();
+
+            // calculate reward
+            // prettier-ignore
+            uint256 totalReward = toTrustToken(
+                interest
+                    .mul(distributor.remaining())
+                    .mul(rewardMultiplier)
+                    .div(distributor.amount())
+            );
+
+            uint256 ratersReward = totalReward.mul(ratersRewardFactor).div(10000);
+            loans[id].reward = ratersReward;
+            if (loans[id].reward > 0) {
+                distributor.distribute(totalReward);
+                trustToken.transfer(address(stakedTrustToken), totalReward.sub(ratersReward));
+            }
+        }
+        _;
+    }
+
+    /**
+     * @dev Claim TRU rewards for voters
+     * - Only can claim TRU rewards for funded loans
+     * - Voters can claim a portion of their total rewards over time
+     * - Claimed automatically when a user withdraws stake
+     *
+     * chi = (TRU remaining in distributor) / (Total TRU allocated for distribution)
+     * interest = (loan APY * term * principal)
+     * R = Total Reward = (interest * chi)
+     * R is distributed to voters based on their proportion of votes/total_votes
+     *
+     * Claimable reward = R x (current time / total time)
+     *      * (account TRU staked / total TRU staked) - (amount claimed)
+     *
+     * @param id Loan ID
+     * @param voter Voter account
+     */
+    function claim(address id, address voter) public override onlyFundedLoans(id) calculateTotalReward(id) {
+        uint256 claimableRewards = claimable(id, voter);
+
+        if (claimableRewards > 0) {
+            // track amount of claimed tokens
+            loans[id].claimed[voter] = loans[id].claimed[voter].add(claimableRewards);
+            // transfer tokens
+            require(trustToken.transfer(voter, claimableRewards));
+            emit Claimed(id, voter, claimableRewards);
+        }
+    }
+
+    function claimed(address id, address voter) external view returns (uint256) {
+        return loans[id].claimed[voter];
+    }
+
+    function claimable(address id, address voter) public view returns (uint256) {
```

Comment:
We also can simplify the claim function since we will only ever claim once. We can greatly simplify a lot of the claiming logic.

Reply:
We still need claimable function though. Would be nice to display in the UI what will you get after withdrawing.